proposals/0055-transact-without-require-weight-at-most.md

```diff
@@ -0,0 +1,65 @@
+---
+Title: Transact without `require_weight_at_most`
+Number: 55
+Status: Draft
+Version: 0
+Authors:
+ - Adrian Catangiu
+ - Francisco Aguirre
+Created: 2024-04-25
+Impact: High
+Requires:
+Replaces:
+---
+
+## Summary
+
+The Transact instruction shouldn't need the `require_weight_at_most` parameter, since the decoder could just get the weight of the call from its dispatch info.
+
+## Motivation
+
+The UX of using `Transact` is not great, and one part of the problem is guesstimating this `require_weight_at_most`.
+We've seen multiple `Transact`s on-chain failures caused by the "incorrect" use or the parameter. In practice,
+this parameter only adds UX overhead. The vast majority of use cases fall in one of two categories:
+1. Unpaid execution of Transacts - in these cases the `require_weight_at_most` is not really useful, caller doesn't 
+have to pay for it, and on the call site it either fits the block or not;
+2. Paid execution of _single_ Transact - the weight to be spent by the Transact is already covered by the `BuyExecution`
+weight limit parameter.
+
+We've had multiple OpenGov `root/whitelisted_caller` proposals initiated by core-devs completely or partially fail
+because of incorrect configuration of `require_weight_at_most` parameter. This is a strong indication that the
+instruction is hard to use.
+
+## Specification
+
+The proposal is simple: change `Transact` instruction:
+
+```diff
+- Transact { origin_kind: OriginKind, require_weight_at_most: Weight, call: DoubleEncoded<Call> },
++ Transact { origin_kind: OriginKind, maybe_weight_limit: Option<WeightLimit>, call: DoubleEncoded<Call> },
+```
+
+With this new API, users who do not need to artificially limit the maximum weight used by the inner `call`,
+can pass `maybe_weight_limit: None`; while those who need to do it, still can.
```

Comment:
Who are the ones that still want to do this?

Reply:
I don't know, maybe you want to run an unpaid execution XCM program that does a bunch of things with a `Transact` somewhere in the beginning and you need to make sure it all runs, not just partially.

The `maybe_weight_limit` could be a way to ensure that the whole thing fits in a block on destination?

Not convinced we actually need it, but starting this RFC conservatively :)